storage/src/sparse_merkle/internal_node.cpp

```diff
@@ -0,0 +1,171 @@
+    stored_child = child;
+    update_hashes(index, child.key.version());
+    return result;
+  }
+  return std::nullopt;
+}
+
+BatchedInternalNode::InsertResult BatchedInternalNode::split_node(size_t index, LeafChild& child, size_t depth) {
+  auto version = child.key.version();
+  auto stored_leaf_child = std::get<LeafChild>(children_.at(index).value());
+
+  // Create a new InternalChild at index.
+  children_.at(index).emplace(InternalChild{Hash(), version});
+
+  auto h = height(index);
+  if (h == 0) {
+    // We've reached the leaf of this node. Tell the caller to create new
+    // BatchedInternalNodes and insert the previously stored leaf as well as the
+    // leaf it was attempting to insert.
+    //
+    // We purposefully don't call `update_hashes`, since we don't know the true
+    // value of the hash of the child until it gets inserted. The caller will fix
+    // this BatchedInternalNode appropriately when the hash is known.
+    return BatchedInternalNode::CreateNewBatchedInternalNodes{stored_leaf_child};
+  }
+
+  // How many prefix bits do the two leaf key hashes have in common?
+  auto bits_in_common = stored_leaf_child.key.hash().bits_in_common(child.key.hash(), depth);
+
+  // Fill in this node with any necessary internal children.
+  size_t add_internal_children_until_height = 0;
+  if (bits_in_common < Nibble::SIZE_IN_BITS) {
+    add_internal_children_until_height = Nibble::SIZE_IN_BITS - bits_in_common;
+  }
+  Nibble child_key = child.key.hash().get_nibble(depth);
+  for (size_t i = h - 1; i >= add_internal_children_until_height && i != SIZE_MAX; i--) {
+    if (child_key.get_bit(i)) {
+      index = right_child_index(index);
+    } else {
+      index = left_child_index(index);
+    }
+    children_.at(index).emplace(InternalChild{Hash(), version});
+  }
+
+  // Is there room to insert the 2 children in this BatchedInternalNode?
+  if (bits_in_common < Nibble::SIZE_IN_BITS) {
+    size_t new_child_index = 0;
+    size_t stored_child_index = 0;
+    if (child_key.get_bit(bits_in_common)) {
+      new_child_index = right_child_index(index);
+      stored_child_index = left_child_index(index);
+    } else {
+      new_child_index = left_child_index(index);
+      stored_child_index = right_child_index(index);
+    }
+    children_[stored_child_index] = stored_leaf_child;
+    children_[new_child_index] = child;
+    update_hashes(new_child_index, version);
+    return BatchedInternalNode::InsertComplete{};
+  }
+
+  // We've reached the leaf of this node. Tell the caller to create new
+  // BatchedInternalNodes and insert the previously stored leaf as well as the
+  // leaf it was attempting to insert.
+  //
+  // We purposefully don't call `update_hashes`, since we don't know the true
+  // value of the hash of the child until it gets inserted. The caller will fix
+  // this BatchedInternalNode appropriately when the hash is known.
+  return BatchedInternalNode::CreateNewBatchedInternalNodes{stored_leaf_child};
+}
```

Comment:
The function is too big.

Reply:
I broke out the two large clauses at the bottom into their own functions. I think this does make it a bit clearer to see what's going on.